Refactor `ToggleButtons` (remove `RawMaterialButton`)

TahaTesser opened the pull request:
fixes https://github.com/flutter/flutter/issues/99085
fixes https://github.com/flutter/flutter/issues/97302

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

const List<Widget> icons = <Widget>[
  Icon(Icons.ac_unit),
  Icon(Icons.call),
  Icon(Icons.cake),
];

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Material App',
      home: ToggleButtonsSample(),
    );
  }
}

class ToggleButtonsSample extends StatelessWidget {
  ToggleButtonsSample({Key? key}) : super(key: key);
  final ValueNotifier<int> _toggleValue = ValueNotifier<int>(0);

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;

    return Scaffold(
      appBar: AppBar(
        title: const Text('ToggleButtons Sample'),
      ),
      body: Center(
        child: ValueListenableBuilder(
          valueListenable: _toggleValue,
          builder: (BuildContext context, int value, Widget? child) {
            return ToggleButtons(
              onPressed: (int newIndex) => _toggleValue.value = newIndex,
              borderRadius: const BorderRadius.all(Radius.circular(8)),
              selectedBorderColor: colorScheme.primary,
              selectedColor: colorScheme.primary,
              isSelected: <bool>[
                _toggleValue.value == 0,
                _toggleValue.value == 1,
                _toggleValue.value == 2,
              ],
              children: icons,
            );
          },
        ),
      ),
    );
  }
}
``` 
	
</details>


### Description
Here this PR refactors `ToggleButtons` and use `ButtonStyleButton` to style `ToggleButtons` 

highlight color not being translated to ElevatedButton (`ButtonStyleButton` sets highlight color to be [transparent](https://github.com/flutter/flutter/blob/6e2d31459d08d27e6076b087244b8bc7811687cf/packages/flutter/lib/src/material/button_style_button.dart#L368)).



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Piinks reviewed (commented):
NICE - all of the existing golden file tests pass! 🎉  (There are only two, but still, that's great)

Comparing this to the same refactor you are doing for FAB (https://github.com/flutter/flutter/pull/99753)

- FAB is not becoming a `ButtonStyleButton`, but here it looks like `ToggleButtons` are.
- FAB is not going to have a `ButtonStyle style` parameter like here, instead deferring to its numerated properties and inherited theme to create a ButtonStyle internally. ToggleButtons here allow for style to be set OR individual properties + theming.

I wonder if the FAB approach is cleaner, the style property here on ToggleButtons might not be necessary, but again I will defer to @HansMuller on general design since he wrote the model. 

All of this work in FAB & Toggle is really great progress though. Thanks!

HansMuller commented:
We should make sure that this PR also fixes https://github.com/flutter/flutter/issues/97302

HansMuller commented:
I agree [with Kate](https://github.com/flutter/flutter/pull/99493#pullrequestreview-906533911), this PR shouldn't include the extra step of adding a ButtonStyle `style` parameter to ToggleButtons; we should leave that for a separate proposal/PR.

TahaTesser committed:
Remove buttonstyle theme property

TahaTesser committed:
Remove internal _Inputpadding and use TextButton tapTargetSize

TahaTesser commented:
> We should make sure that this PR also fixes https://github.com/flutter/flutter/issues/97302

This is now handled by `TextButton` widget in this refactor, can you please take a look

### Mobile

![Screenshot_1647249245](https://user-images.githubusercontent.com/48603081/158218908-f4c016ee-b03f-4ba1-882e-4a09af370644.png)

### Desktop (Web)
![Screenshot 2022-03-14 111522](https://user-images.githubusercontent.com/48603081/158218963-5bb66218-2376-4d8b-8c7e-4971a291c347.png)

HansMuller commented:
@TahaTesser - if this PR fixes #97302 it should say so in the PR's description and it should include a regression test (marked as a regression test for #97302 with a comment).

HansMuller commented:
@TahaTesser - the screenshots in https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 imply that the background color for the ToggleButtons isn't clipped to the ToggleButtons' shape?

TahaTesser commented:
@HansMuller 
Thanks so much for your time,

> @TahaTesser - the screenshots in https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 imply that the background color for the ToggleButtons isn't clipped to the ToggleButtons' shape?

This is just a parent container with color from https://github.com/flutter/flutter/issues/9730. 

ClipRRect still clipping the toggle buttons in this refactor
https://github.com/flutter/flutter/pull/99493/files#diff-153bdbb7a6739dd071bbb7b8230e3884dcb6cedbc21a78ef82c0e6a8166bb220R731

HansMuller commented:
@TahaTesser - Good!

TahaTesser committed:
MaterialTapTargetSize regression test

TahaTesser committed:
Remove indent

TahaTesser commented:
@HansMuller 
Added a regression test with a link to the issue

I am currently investigating this behavior, just noticed this, this seems to be only happening in the vertical direction.
| Preview | 
| --------------- | 
| <img src="https://user-images.githubusercontent.com/48603081/158273202-53098392-fe2a-4684-a8d3-e14625975a2c.gif" height="450" />	|

TahaTesser commented:
@HansMuller 
During refactoring noticed, there is no interactive example that showcases different configurations for `ToggleButtons`

Filed https://github.com/flutter/flutter/pull/100124, this will land after this PR (tester finders look for `TextButton` instead of `RawMaterialButton`).

HansMuller reviewed (commented):
This is looking good however I don't think the style parameter belongs in this PR; see https://github.com/flutter/flutter/pull/99493#issuecomment-1064643880

TahaTesser commented:
> This is looking good however I don't think the style parameter belongs in this PR; see https://github.com/flutter/flutter/pull/99493#issuecomment-1064643880

My apologies I missed removing that parameter, done.

TahaTesser commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -668,33 +669,87 @@ class ToggleButtons extends StatelessWidget {
+          child: TextButton(
+            focusNode: focusNodes != null ? focusNodes![index] : null,
+            style: ButtonStyle(
+              backgroundColor: MaterialStateProperty.all<Color?>(effectiveFillColor),
```
Thanks! Removed nullable except `mouseCursor`  since it is using a direct nullable widget parameter.

TahaTesser commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -668,33 +669,87 @@ class ToggleButtons extends StatelessWidget {
+          if (isSelected[index] && onPressed != null) MaterialState.selected,
+          if (onPressed == null) MaterialState.disabled,
+      };
+      final Color effectiveFillColor = _ResolveFillColor(fillColor
```
I merged the two and it is provided directly to the` ButtonStyle` now since the conditioning is inside `_ResolveFillColor` now.

TahaTesser commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -668,33 +669,87 @@ class ToggleButtons extends StatelessWidget {
+                hoverColor: hoverColor ?? toggleButtonsTheme.hoverColor,
+                splashColor: splashColor ?? toggleButtonsTheme.splashColor,
+              ),
+              elevation: MaterialStateProperty.all<double>(0),
```
Removed these redundant style properties.

Piinks reviewed (commented):
@TahaTesser is this ready for another review? Or is it waiting on additional changes like the FAB one?

TahaTesser commented:
> @TahaTesser is this ready for another review? Or is it waiting on additional changes like the FAB one?

Hey @Piinks 
This needs another review. 

This `Diagnosticable` issue from FAB one doesn't affect this PR, all tests are passing

HansMuller commented:
@TahaTesser - I haven't tested or even compiled the code I've suggested for the _ForegroundColor and _BackgroundColor. If they're difficult let me know and I'll fix them.

TahaTesser commented:
> @TahaTesser - I haven't tested or even compiled the code I've suggested for the _ForegroundColor and _BackgroundColor. If they're difficult let me know and I'll fix them.

This is cleaner, thanks! 
I made slight adjustments and added these classes including the local ColorScheme variable as suggested and removed the redundant code due to these classes.

HansMuller commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -668,33 +669,69 @@ class ToggleButtons extends StatelessWidget {
+            focusNode: focusNodes != null ? focusNodes![index] : null,
+            style: ButtonStyle(
+              backgroundColor: _BackgroundColor(
+                onPressed != null && isSelected[index],
```
If you define a isSelected variable above, then these constructors and their positional parameters should fit neatly on one line. The same `onPressed != null && isSelected[index]` appears in the overlayColor value too; definitely wouldn't hurt to factor it out.

TahaTesser replied:
Great idea, I added a local variable and made those into a single line 

Also added regression for https://github.com/flutter/flutter/issues/97302 (I added it earlier too but somehow missed it during buttonstyle revert).  This test would be able to close the issue :)

HansMuller reviewed (approved):
LGTM

Piinks commented:
I just got an alert that this failed 1 internal test, but the test itself needs to be fixed simultaneously with this change landing. It is possible to do, let's just hold on landing it until I have the fix ready to land. I'll ping back once it's ready.

Piinks commented:
Update: The fix is out for review, awaiting approval.

Piinks commented:
@TahaTesser, can you update your branch with the latest from master? The google testing shard is stuck. I have the fix ready, so once it is rebased we can land it.

math1man commented:
Hey, I think this change is a incorrect. https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 does not look like the correct toggle button styling on Mobile.

HansMuller commented:
@math1man - can you explain what aspect of the component doesn't look correct?

math1man commented:
It's too tall.

HansMuller commented:
@TahaTesser - can you investigate? AFAICT we have not changed the geometry of ToggleButtons, however perhaps we've always had this wrong. I don't believe https://m3.material.io/ covers it (yet).

Piinks commented:
The failure appears related to https://github.com/flutter/flutter/issues/97302, which this closed. Rather than allow for a height of 32, it seems like this enforced a height of 48 - IIRC that is the minimum required size for interactive elements. To reproduce this, the ToggleButtons.constraints were set with a min/max height of 32, but did not adhere to that.

TahaTesser commented:
@HansMuller 

I also found this in the Material Design components app, I will look into and test it in an Android native app.

| Preview | 
| --------------- | 
| <img src="https://user-images.githubusercontent.com/48603081/162292942-78361fe0-ac56-4f39-b813-614fcb051a0b.jpg" height="450" />	|


Perhaps there should Android equivalent of https://github.com/flutter/platform_tests/tree/master/ios_widget_catalog_compare

TahaTesser commented:
@Piinks 
Thank you for reverting

HansMuller commented:
If a PR is created that relands the ToggleButton refactor, we'll need some additional regression tests to verify that the buttons handle input within 48 vertical without the button's outline occupying 48.
